### PR TITLE
Adding support for Oracle Cloud

### PR DESCRIPTION
**Pull request by jhorvath**
Adding new module with support for Oracle Cloud. For now it is able to browser Tenancy and Compartments. In selected Compartment it shows Autonomous Databases and allow to create and download Database Wallet.

### CODE REVIEW

File: enterprise/cloud.oracle/external/binaries-list
Comment (sdedic) [inline enterprise/cloud.oracle/external/binaries-list:28]: FYI - some of the libraries (commons-lang, commons-codec) also appears as wrapped libraries and/or in other modules. Maybe we could upgrade the shipped wrapped libraries or share the libraries between modules ?

jackson seems also part of other enterprise module, bouncycastle has its own wrapped already.

File: .gitignore
Comment (sdedic) [inline .gitignore:22]: Unnecessary, libs are downloaded to `external`

File: ide/server/src/org/netbeans/api/server/CloudProvider.java
Comment (sdedic) [inline ide/server/src/org/netbeans/api/server/CloudProvider.java:33]: Add `apichanges.xml` entry, bump `ide/server` spec version, depend on the new version in `cloud.oracle`

File: enterprise/cloud.oracle/external/binaries-list
Comment (JaroslavTulach) [reply enterprise/cloud.oracle/external/binaries-list:28]: +1 we want to share the libraries, if possible.

File: enterprise/cloud.oracle/build.xml
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/build.xml:1]: No license, but no failed build. Strange.

File: enterprise/cloud.oracle/nbproject/project.properties
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/nbproject/project.properties:31]: I believe these libraries will need `licenseinfo.xml` in the module - CCing @matthiasblaesing  - again I am surprised, no build fails.

File: enterprise/cloud.oracle/test/unit/src/org/netbeans/modules/cloud/oracle/DownloadWalletDialogTest.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/test/unit/src/org/netbeans/modules/cloud/oracle/DownloadWalletDialogTest.java:37]: Great test for the UI!

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OpenServiceConsoleAction.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OpenServiceConsoleAction.java:36]: A new category?

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DownloadWalletAction.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DownloadWalletAction.java:41]: Shouldn't the ID better be derived (prefixed) from code name base of the module?

File: enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIManager.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIManager.java:88]: I don't like `Optional`. Can't you return `null`?

File: enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIManager.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIManager.java:154]: Where the downloaded wallet is stored?

File: enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/DatabaseItem.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/DatabaseItem.java:25]: API, really? For whom? 

One thought: API modules should better be autoload - then they can't have UI registration. E.g. contrary to current state - however it all depends on the actual use-case.

File: enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIItem.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/api/cloud/oracle/OCIItem.java:36]: Subclassable by public!?

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DatabaseNode.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DatabaseNode.java:45]: The path for actions is an API and needs a note in `arch.xml`. How did you decided on the `could/oracle/db` path?

File: ide/server/src/org/netbeans/api/server/CloudProvider.java
Comment (JaroslavTulach) [reply ide/server/src/org/netbeans/api/server/CloudProvider.java:33]: plus consider rewrite of other registraitons to use this new annotation.

**Review by JaroslavTulach — COMMENTED** [review]
Modify `nbbuild/cluster.properties` to include this module. Modify `nbbuild/build.properties` to include the new Javadoc. Hopefully then the module becomes subject to license & co. checks.

File: nbbuild/antsrc/org/netbeans/nbbuild/extlibs/ignored-overlaps
Comment (JaroslavTulach) [inline nbbuild/antsrc/org/netbeans/nbbuild/extlibs/ignored-overlaps:128]: Are these overlaps OK? I am inclined to believe they are. Having clashes with gradle/maven distributions seems OK to me...

File: platform/o.apache.commons.lang3/build.xml
Comment (JaroslavTulach) [inline platform/o.apache.commons.lang3/build.xml:26]: I guess this follows a pattern copied from somewhere else, right?

**Review by JaroslavTulach — APPROVED** [review]
Wrapping libraries into modules, updating versions of some libraries - looks like good job. Creating and using the `@CloudProvider` looks fine as well. I don't see problems from architecture perspective.

File: enterprise/cloud.oracle/external/j2objc-annotations-1.3-license.txt
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/external/j2objc-annotations-1.3-license.txt:2]: Wow. Why do we need this?

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DatabaseNode.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/DatabaseNode.java:44]: This is a `<api group="layer" .../>` and it should be documented in `arch.xml` document. It should be noted that there is an instance of `OCIItem` in the lookup - e.g. one can write actions that react to `OCIItem`.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCIItem.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCIItem.java:26]: The `OCIItem` would then be the API of this module... but right now it would be _private API_ as the class isn't exposed.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/TenancyNode.java
Comment (JaroslavTulach) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/TenancyNode.java:82]: Looks like empty implementation.

**Review by JaroslavTulach — APPROVED** [review]
What's the license for `enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/resources/orcl.png`? There should be `licenseinfo.xml` file specifying that. Btw. are you sure the compartment and database `.svg` files are really licensed under Apache license?

**Comment by JaroslavTulach** [comment]
Except the image licenses (which Jan Horváth will clarify) I consider the integration solid. Any other review comments? Otherwise I plan to merge this week.

File: enterprise/websvc.restlib/nbproject/project.xml
Comment (matthiasblaesing) [inline enterprise/websvc.restlib/nbproject/project.xml:163]: These exports look wrong to me. Shouldn't the module wrap

**Review by matthiasblaesing — CHANGES_REQUESTED** [review]
As @JaroslavTulach already pointed out the PNG needs a license to it `enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/resources/orcl.png`.

The rest of the changes looks ok to me (only eyeballed!).

For the library updates, this duplicates/overrides #2287. It would be good if the bcpg module would then also be updated.

In general this PR needs cleanup, we should not merge this with 28 commits. After review and before merge these should either be fully squashed into a single commit or squashed based on logical units.

File: enterprise/cloud.oracle/external/j2objc-annotations-1.3-license.txt
Comment (jhorvath) [reply enterprise/cloud.oracle/external/j2objc-annotations-1.3-license.txt:2]: This is one of the transitive dependencies of OCI libraries. I tried to remove it and fortunately it works. Removing.

File: platform/o.apache.commons.lang3/build.xml
Comment (jhorvath) [reply platform/o.apache.commons.lang3/build.xml:26]: Copied from another apache commons lib module.

File: nbbuild/antsrc/org/netbeans/nbbuild/extlibs/ignored-overlaps
Comment (jhorvath) [reply nbbuild/antsrc/org/netbeans/nbbuild/extlibs/ignored-overlaps]: Comment was updated with more accurate - _Maven and Gradle are self-contained distributions - ignoring overlaps._
Thanks Jardo.

Comment (jhorvath) [reply nbbuild/antsrc/org/netbeans/nbbuild/extlibs/ignored-overlaps]: I tries to update `bcpg`, but `CachingPreventsLoadingOfModuleManifestsTest.testDontLoadManifests()` started failing. As this PR is already too big, I would solve the update separately.

**Comment by sdedic** [comment]
All green two approvals, changes squashed. Merging.

**Comment by neilcsmith-net** [comment]
Thanks @sdedic Please make sure to check and add the milestone when merging.

**Pull request merged**
